Replace unsafe signal handlers with signalfd mechanism.

alindima opened the pull request:
# Reason for This PR

https://github.com/firecracker-microvm/firecracker/issues/2452 

Resources:
https://man7.org/linux/man-pages/man2/signalfd.2.html
https://man7.org/linux/man-pages/man7/signal.7.html

## Description of Changes

View commit messages

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

acatangiu commented:
This being an RFC I guess I should have added that I agree with the approach and the implementation and I suggest we promote this to a full PR 👍

sandreim commented on src/vmm/src/signal_handler.rs:
```diff
@@ -128,162 +156,95 @@ generate_handler!(
+        SIGSYS => sigsys_handler(siginfo),
+        SIGHUP => sighup_handler(siginfo),
+        SIGILL => sigill_handler(siginfo),
+        // This should never be reached since we'll only receive the signals in the signalfd mask.
```
We should add an error! message here, to document the unexpected FC exit in the log.

sandreim commented on src/vmm/src/signal_handler.rs:
```diff
@@ -128,162 +156,95 @@ generate_handler!(
-            assert!(METRICS.seccomp.num_faults.count() >= 1);
+        // Even though we currently kill the process during signal handling, we should in theory
+        // read from the signalfd all the signals for future-proofing.
+        while self.signal_fd.read(&mut read_buf[..]).unwrap() == siginfo_size {
```
Can the read fail ?

sandreim commented on src/vmm/src/signal_handler.rs:
```diff
@@ -128,162 +156,95 @@ generate_handler!(
+        // Even though we currently kill the process during signal handling, we should in theory
+        // read from the signalfd all the signals for future-proofing.
+        while self.signal_fd.read(&mut read_buf[..]).unwrap() == siginfo_size {
+            let siginfo: signalfd_siginfo = unsafe { ptr::read(read_buf.as_ptr() as *const _) };
```
Is this safe ?

alindima replied:
In the latest version, I replaced this unsafe call with a call to the `ByteValued::from_slice()` which returns an option, so we are now good in terms of error handling. If the read is not possible, we should get a `None` and we log the error.

alindima commented on tests/integration_tests/functional/test_signals.py:
```diff
@@ -139,7 +142,7 @@ def test_handled_signals(test_microvm_with_ssh, network_config):
     assert int(stdout.read()) == 2
 
     # We have a handler installed for this signal.
-    os.kill(firecracker_pid, SIGRTMIN+1)
```
Hardcoded the 35 since the `SIGRTMIN` python reports is not correct for musl. It's likely the one for glibc, and we shouldn't rely on it.

alindima commented:
Implemented unit tests. PR is no longer RFC

alindima commented on src/vmm/src/vstate/vcpu/mod.rs:
```diff
@@ -241,6 +242,8 @@ impl Vcpu {
         let vcpu_thread = thread::Builder::new()
             .name(format!("fc_vcpu {}", self.kvm_vcpu.index))
             .spawn(move || {
+                mask_handled_signals().expect("Unable to install signal mask on VCPU thread.");
```
Very nice catch. VCPU threads inherit the mask from the VMM thread, since they are created after the signalfd is registered.

The API thread is the one who inherits empty signal masks, since at the time it is launched, the `signalfd` was not yet created. And we don't want to mask the signals unless we have the signalfd mechanism in place (to not lose any signals data).

alindima replied:
Removed this

alindima commented on src/vmm/src/signal_handler.rs:
```diff
@@ -1,38 +1,66 @@
+    signal_fd: File,
+}
+
 macro_rules! generate_handler {
```
This is true. I will try to find a prettier way of defining these functions, since there is no longer a need for a macro.

alindima replied:
Done. PTAL @serban300

alindima commented on src/vmm/src/signal_handler.rs:
```diff
@@ -128,162 +156,95 @@ generate_handler!(
-                syscall(libc::SYS_kill, process::id(), SIGSEGV);
-            }
+/// Handle the signal according to its signal number.
+fn handle_signal(siginfo: signalfd_siginfo) {
```
I was on the fence about this, since it does not need anything from the SignalManager state.
I agree however that it is cleaner to have it in the same scope

alindima replied:
done

alindima commented:
Addressed all comments except for one which still needs work: replacing the `generate_handler` macro with something prettier

alindima commented:
Replaced the unsafe call to `ptr::read` with a call to the `ByteValued` trait.
Also removed the signal handler generator macros and moved the handler logic in the `SignalManager::handle_signal()` method

serban300 commented on src/vmm/src/signal_handler.rs:
```diff
@@ -229,23 +136,60 @@ impl SignalManager {
+        }
+
+        // Increment the right metrics.
+        match si_signo {
```
I would rather have a function that does this match and returns a tuple with the metric and exit code associated with each signal.

alindima replied:
done

serban300 commented on src/vmm/src/signal_handler.rs:
```diff
@@ -1,168 +1,254 @@
+        loop {
+            match self.signal_fd.read(&mut read_buf[..]) {
+                Ok(size) if size == siginfo_size => {
+                    match SignalInfoWrapper::from_slice(&read_buf) {
```
I don't think this is worth the added complexity introduced by `SignalInfoWrapper`. Personally I would just read. It should be safe.

alindima replied:
As we discussed offline, there is not much added complexity and the `ByteValued` approach has some clear advantages:

 - does not copy the buffer, it uses the same memory as the byte slice, as opposed to the `ptr::read`
 - has some extra safety checks: checks that the data is correctly aligned and checks the length of the byte slice.

alindima commented:
Addressed all comments. PTAL @acatangiu @sandreim

georgepisaltu reviewed (approved):
🚀
